codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -211,134 +263,245 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                onHttp2StreamStateChanged(ctx, stream2);
+            }
         }
 
         @Override
-        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
-            ctx.fireChannelRead(new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData.retain()));
+        public void onStreamHalfClosed(Http2Stream stream) {
+            DefaultHttp2FrameStream stream2 = stream.getProperty(streamKey);
+            if (stream2 != null) {
+                onHttp2StreamStateChanged(ctx, stream2);
+            }
         }
     }
 
-    private static final class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
-        InternalHttp2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
-                                       Http2Settings initialSettings) {
-            super(decoder, encoder, initialSettings);
+    @Override
+    protected final void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Http2Exception http2Ex) {
+        ctx.fireExceptionCaught(cause);
+    }
+
+    /**
+     * Exceptions for unknown streams, that is streams that have no {@link Http2FrameStream} object attached
+     * are simply logged and replied to by sending a RST_STREAM frame.
+     */
+    @Override
+    protected final void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                                 Http2Exception.StreamException streamException) {
+        int streamId = streamException.streamId();
+        Http2Stream connectionStream = connection().stream(streamId);
+        if (connectionStream == null) {
+            onHttp2UnknownStreamError(ctx, cause, streamException);
+            // Write a RST_STREAM
+            super.onStreamError(ctx, cause, streamException);
+            return;
         }
 
-        @Override
-        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
-                                     Http2Exception.StreamException http2Ex) {
-            try {
-                Http2Stream stream = connection().stream(http2Ex.streamId());
-                if (stream == null) {
-                    return;
-                }
-                ctx.fireExceptionCaught(http2Ex);
-            } finally {
-                super.onStreamError(ctx, cause, http2Ex);
-            }
+        Http2FrameStream stream = connectionStream.getProperty(streamKey);
+        if (stream == null) {
+            LOG.warn("Stream exception thrown without stream object attached.", cause);
+            // Write a RST_STREAM
+            super.onStreamError(ctx, cause, streamException);
+            return;
         }
+
+        onHttp2FrameStreamException(ctx, new Http2FrameStreamException(stream, streamException.error(), cause));
     }
 
-    private static final class FrameListener extends Http2FrameAdapter {
+    void onHttp2UnknownStreamError(@SuppressWarnings("unused") ChannelHandlerContext ctx, Throwable cause,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)